content/en/blog/_posts/2025-04-21-kubernetes-network-evolution/index.md

```diff
@@ -0,0 +1,71 @@
+
+## The Road to Declarative Networking: Key Milestones
+
+With CNI established as the foundation, the Kubernetes community began to work on more advanced networking requirements, particularly the need for a more declarative networking support. This journey was marked by several key initiatives and discussions:
+
+### The Network Plumbing Working Group (NPWG)
+
+The Network Plumbing Working Group (NPWG) was formed at Kubecon NA '17 in Austin to initially work towards implementing a common standard addressing how one might attach multiple networks to pods in Kubernetes. The work of the NPWG has culminated in a [standard Multi-Network Custom Resource Definition (CRD)](https://github.com/k8snetworkplumbingwg/multi-net-spec). The [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni) is the reference implementation of this multi-network definition, demonstrating how pods can have multiple network interfaces.
+
+### KEP-3698 (Multi-Network): Reviving the Discussion
+
+The first significant step towards directly addressing multi-network requirements within the core Kubernetes project was the introduction of [KEP-3698 (Multi-Network)](https://github.com/kubernetes/enhancements/pull/3700). This Kubernetes Enhancement Proposal aimed to define user stories and requirements for multi-network functionality in Kubernetes. However, early feedback indicated that the KEP was perceived as too abstract and broad, attempting to encompass a wide range of use cases, from simply attaching multiple interfaces to more complex network virtualization scenarios.
+
+A core goal of KEP-3698 was to introduce an API object to Kubernetes describing networks that Pods can attach to and to evolve the current Kubernetes networking model to support multiple networks in a backward-compatible way. The initial focus was on defining what users wanted to achieve with multi-networking, such as connecting Pods to each other in a manner defined by the network object, and exposing basic network interface information for each attachment.
+
+Discussions around the KEP revealed challenges related to integrating multi-networking in existing Kubernetes architecture. For instance, there were questions about how network probes would function on non-default networks (problem that originated a new KEP [https://github.com/kubernetes/enhancements/pull/4558](https://github.com/kubernetes/enhancements/pull/4558)) and the implications for other core functionalities like DNS and network policies in a multi-network environment. KEP-3698 did an important work on clarifying the end user multi-networking requirements. This lack of defined semantics and standard implementations, coupled with the anticipated need to rewrite existing Kubernetes components, ultimately led the community to explore more feasible alternative approaches.
```

Comment:
KEP-4558 was initially primarily about the NetworkPolicy hole and the SSRF problem. Multi-Network only got dragged in as part of the "while we're here what else can we fix".

Reply:
I kind of assumed that @tssurya was working it because of multinetwork, since the probes issue was open since 2021 https://github.com/kubernetes/kubernetes/issues/102613 .

Reply:
I don't mind taking up that multi-network part as well :D . It might help justify my time upstream for that (downstream :D selflishly we could _try to put that also as one of the main goals_

Reply:
then I misunderstood, and I should fix the text as Dan suggests

Reply:
(FTR, I started working on the KEP because I didn't want AdminNetworkPolicy to need the same "hole" for kubelet probes as NetworkPolicy has.)

Reply:
then I should add that too